### PR TITLE
gpt-auto: remove directory check for ESP mount

### PR DESCRIPTION
**Pull request by NekkoDroid**
According to the [Boot Loader Specification](https://uapi-group.org/specifications/specs/boot_loader_specification/#mount-points) `ESP` should when possible be mounted to `/boot/` unless `XBOOTLDR` is mounted there.

The way it currently is, `ESP` is always mounted to `/efi/` when booting with an empty root partition, which caused a problem since I somewhat relied on the boot entry (UKI) being in `/boot/` with and without a `XBOOTLDR` partition.

So just remove the directory check to follow the spec more closely, but still keep the fallback to `/efi/` for a busy directory to mostly remain backwards compatible.

### CODE REVIEW

**Comment by YHNdnzj** [comment]
cc @keszybz

**Comment by poettering** [comment]
Hmm, so, this touches the rules established by: 6a488fa7cce8124fa885adf8a2f31363fe62f636

I think this change makes sense, hence green labelling it.

@keszybz you might want to have a look too, as you defined original rules for this.

File: src/gpt-auto-generator/gpt-auto-generator.c [hunk 1]
```diff
@@ -538,28 +538,28 @@ static int add_partition_esp(DissectedPartition *p, bool has_xbootldr) {
         if (r > 0)
                 return 0;
 
-        /* If /boot/ is present, unused, and empty, we'll take that.
-         * Otherwise, if /efi/ is unused and empty (or missing), we'll take that.
+        /* Only mount to /efi/ if we have a XBOOTLDR partition or /efi/ already exists.
+         * Otherwise, mount to /boot/ if unused and empty (or missing) or if /efi/ is busy.
          * Otherwise, we do nothing. */
```
Comment (YHNdnzj) [inline src/gpt-auto-generator/gpt-auto-generator.c]: Well, this means we'd no longer create `/efi/` even if `/boot/` is in use (not by XBOOTLDR)? That sounds problematic to me. We should always fall back to `/efi/`.

**Comment by NekkoDroid** [comment]
This should now address @YHNdnzj concerns. Effectivly this now just inverts the directory check from "boot exists" to "efi not exists", which should also work for my usecase

File: src/gpt-auto-generator/gpt-auto-generator.c [hunk 2]
```diff
@@ -538,10 +538,10 @@ static int add_partition_esp(DissectedPartition *p, bool has_xbootldr) {
         if (r > 0)
                 return 0;
 
-        /* If /boot/ is present, unused, and empty, we'll take that.
+        /* Only attempt to mount to /boot/ if we there is no XBOOTLDR partition and /efi/ doesn't already exists.
```
Comment (behrmann) [inline src/gpt-auto-generator/gpt-auto-generator.c]:
```suggestion
        /* Only attempt to mount to /boot/ if there is no XBOOTLDR partition and /efi/ doesn't already exists.
```

**Comment by YHNdnzj** [comment]
> This should now address @YHNdnzj concerns. Effectivly this now just inverts the directory check from "boot exists" to "efi not exists", which should also work for my usecase.

hrm, with the new logic `/efi/` is prioritized now? I still don't follow?

If we're changing this, the sane behavior IMO is to create/use `/boot/` unless it's XBOOTLDR or not empty, where `/efi/` is used. Basically it means we shouldn't care about the existence of dirs at all.

**Comment by NekkoDroid** [comment]
> hrm, with the new logic /efi/ is prioritized now? (if it exists we'd always use it) I still don't follow?

Yes, if the directory is explicitly created (be it manually or from a previous mount) I do think it makes sense to mount there, as it was "explicitly" requested. So if we are going to take existence of directories into account that would be the expected behavior.

I have no problem just changing this to only consider if XBOOTLDR exists or not. I probably would prefer that, as that makes the mount location easier to reason about (XBOOTLDR exists? then /efi/, else /boot/). If that is what is wanted I can do that.

**Comment by YHNdnzj** [comment]
6a488fa7cce8124fa885adf8a2f31363fe62f636 changed gpt-auto to try `/boot/` first even if `/efi/` exists. Generally it's better, because even if the user decides to remove XBOOTLDR partition one day `/boot/` would still work, without the need to manually remove `/efi/` dir. Hence I'd prefer the other way around.

> I have no problem just changing this to only consider if XBOOTLDR exists or not.

But do note that this would not suffice. E.g. in my own Arch system `/boot/` resides on `/`, and kernels would be installed into `/boot/` by mkinitcpio, which later gets read and constructed into a UKI that ends up in `/efi/`. Whether the dir is empty needs to be taken into account.

--------------

Ah, the PR is updated before this comment is complete. The new approach looks good to me.

**Review by YHNdnzj — APPROVED** [review]

**Comment by NekkoDroid** [comment]
> But do note that this would not suffice. E.g. in my own Arch system `/boot/` resides on `/`, and kernels would be installed into `/boot/` by mkinitcpio, which later gets read and constructed into a UKI that ends up in `/efi/`. Whether the dir is empty needs to be taken into account.

Interesting. I am somewhat surprised by such a setup, but it does make sense with the current/previous implementation that it ends up like that. I just had `mkinitcpio` output to `/boot` when I had an ESP only system.

**Comment by YHNdnzj** [comment]
(The PR description should be updated BTW)

**Review by yuwata — APPROVED** [review]

**Comment by keszybz** [comment]
After the merge already, but LGTM. Effectively, the only case which is influences by this change is when `/boot` does not exist and we only need one mount point. Previously, we would possibly create `/efi` and mount ESP on it, now we create `/boot` and mount ESP on it. The result is more logical and matches the intent of using `/boot` as the "primary" location for boot entries.

**Comment by nikstur** [comment]
Isn't this a breaking change that isn't a "small non-conflicting feature deemed safe to add in a stable release" ? I'm surprised that it was backported to v256.8. Was that correct?

**Pull request merged**